dvc/objects/db/index.py

```diff
@@ -96,7 +96,7 @@ def __init__(
         self.index_dir = os.path.join(tmp_dir, self.INDEX_DIR, name)
         makedirs(self.index_dir, exist_ok=True)
         self.fs = LocalFileSystem()
-        self.index = Index(self.index_dir)
+        self.index = Index(self.index_dir, disk_pickle_protocol=4)
```

Comment:
@pmrowla, will this break repo if they were using protocol 5 before?

Reply:
Yes, users may need to remove `.dvc/tmp/index` in this case (which is the same potential issue we had when we pinned the version for state https://github.com/iterative/dvc/issues/5563#issuecomment-792728682)

Reply:
Python >=3.8 uses Protocol version 5, so we may as well just don't support compatibility between 3.7 and >=3.8.
What do you think of doing something like:
```suggestion
        self.index = Index(
            self.index_dir,
            disk_pickle_protocol=4 if sys.version_info < (3,8) else 5
        )
```

Reply:
I'm not sure that works here since the reason users are still running into this issue is because they are mixing python versions (between 3.7 and >3.7)

Reply:
Also, the core python [default protocol](https://docs.python.org/3/library/pickle.html#pickle.DEFAULT_PROTOCOL) version in 3.8+ is still 4, the problem for us is that diskcache explicitly uses `pickle.HIGHEST_PROTOCOL` https://github.com/grantjenks/python-diskcache/blob/d55a50ee083784afa9c85e14e41c4a2d132f3111/diskcache/core.py#L60

Reply:
I'm just worried that it may crash for users in the new release unless they remove `.dvc/tmp/index`, although I did not really see any exceptions when I migrated to change it to 4. I guess that works already?

> Also, the core python default protocol version in 3.8+ is still 4, the problem for us is that diskcache explicitly uses pickle.HIGHEST_PROTOCOL grantjenks/python-diskcache@d55a50e/diskcache/core.py#L60

yes, but the highest protocol in 3.8+ is 5.

Reply:
It's not backwards compatible, but it won't raise any exceptions until you actually try to read/write anything from disk

Reply:
For myself, I'd prefer to raise some messages to suggest `.dvc/tmp/index` here.

Reply:
Regarding pickle protocol version, I think we should ensure compatibility across python versions that are supported by DVC, which means we should be enforcing a pinned version wherever we are using pickle (unless at some point in the future Python officially deprecates whatever ver we are using).

In the state/index use case, I don't think there's any practical benefit to using protocol ver 5 for the way we use `diskcache` (my understanding of protocol ver 5 is that it improves performance for serializing very large non-native objects, like numpy arrays and pandas dataframes), but with the way we use diskcache we are really just serializing (relatively) small native dicts.

There is an official backport library for pickle protocol v5, so if/when the state/index pickle performance becomes a real problem for us we can go that route, but I don't think it's a pressing need right now.

Reply:
@pmrowla, I agree with all of your points, I have no issues on the technical side of things.

I'm just worried about potentially breaking things here for Python 3.8+ users in the middle of 2.x version. I'm fine with this as long as we are explicit about this being a breaking change and weigh in both options:

1. By not fixing this, we break compatibility between 3.7 with 3.8+ Python versions. So, if the user used Python 3.8+ version, they cannot use dvc repo again with Python 3.7. 
2. By fixing this, we ensure compatibility between all Python versions by breaking compatibility for once for Python 3.8+ users.

Reply:
@skshetry as a short term workaround we could just catch the protocol errors in state/index, and then add a troubleshooting link w/details suggesting users can try removing the relevant temp dirs (`md5s/links/index`) and then retry the command

Reply:
So after some testing, it turns out that changing to protocol version 4 will not break an existing diskcache cache that was generated using protocol version 5 (as long as you are on a python version that supports v5). It will just rewrite the cache entry using protocol version 4 on the next time that entry is modified.

This test:
```py
@pytest.mark.parametrize("proto_a, proto_b", [(4, 5), (5, 4)])
def test_protocol_error(tmp_dir, proto_a, proto_b):
    with diskcache.Cache(directory=(tmp_dir / "test"), disk_pickle_protocol=proto_a) as cache:
        cache["key"] = ("value1", "value2")
        assert ("value1", "value2")  == cache["key"]
    with diskcache.Cache(directory=(tmp_dir / "test"), disk_pickle_protocol=proto_b) as cache:
        cache["key"] = ("value3", "value4")
        assert ("value3", "value4")  == cache["key"]
```
passes on 3.8, 3.9, 3.10 and fails on 3.7 (since 3.7 cannot read/write protocol version 5)

So this should not break compatibility with existing repos for anyone on python 3.8+ (but generating the troubleshooting message for py 3.7 will still be useful)